javalib/src/main/scala/java/util/Formatter.scala

```diff
@@ -1425,4 +1359,242 @@ object Formatter {
+                token.width = getPositiveInt(format)
+                ApplyWidth
+              } else if ('.' == currentChar) {
+                ApplyPrecision
+              } else {
+                // do not get the next char.
+                format.position(format.position() - 1)
+                ApplyConversionType
+              }
+            }
+
+          case ParserStateMachine.ApplyWidth =>
+            loop {
+              if ('.' == currentChar) ApplyPrecision
+              else {
+                format.position(format.position() - 1)
+                ApplyConversionType
+              }
+            }
+
+          case ParserStateMachine.ApplyPrecision =>
+            if (Character.isDigit(currentChar)) {
+              token.precision = getPositiveInt(format)
+            } else {
+              // the precision is required but not given by the format string.
+              throw new UnknownFormatConversionException(getFormatString(token))
+            }
+            loop(ApplyConversionType)
+
+          case ParserStateMachine.ApplyConversionType =>
+            token.conversion = currentChar
+            if ("bBhHsScCdoxXaAeEgGfn%".indexOf(currentChar) < 0) {
+              val hasArgIdx = token.argIndex != FormatToken.Unset
+              val noWidthOrPrecision = token.width == FormatToken.Unset ||
+                token.precision == FormatToken.Unset
+              /* In Scala.js we used regex matching to determinate if format specifier
+               * is valid, but it's not possible when using current parser.
+               * We provide estimation about the root of problem which should be compliant
+               * with JVM. Checking for EndOfString is done in parser loop.
+               * Based on test cases of `FormatterTest.unknownFormatConversion`
+               */
+              val causedBy = {
+                if (hasArgIdx && noWidthOrPrecision) token.argIndex.toString
+                else token.conversion.toString
+              }
+
+              throw new UnknownFormatConversionException(causedBy)
+            }
+            if (currentChar <= 'Z') {
+              token.setFlag(Flags.UpperCase)
+            }
+            loop(Exit)
+
+          case ParserStateMachine.ApplySuffix =>
+            // Todo `t` | `T` support
+            loop(Exit)
+        }
+      }
+
+      loop(Entry)
+    }
+
+    private def getNextFormatChar(): Char = {
+      if (format.hasRemaining()) format.get()
+      else EOS.toChar
+    }
+
+    private def getFormatString(token: FormatToken): String = {
+      val end = format.position()
+      format.rewind()
+      val formatString =
+        format.subSequence(token.formatStringStartIndex, end).toString
+      format.position(end)
+      formatString
+    }
+
+    private def getPositiveInt(buffer: CharBuffer): Int = {
+      def loop(): Int = {
+        if (buffer.hasRemaining() && Character.isDigit(buffer.get())) {
```

Comment:
`Character.isDigit` is too broad, as it accepts all Unicode characters considered as digits.

We should only recognize `c >= '0' && c <= '9'`.

There are other places that use `Character.isDigit` and which should all be adapted.